Common++/header/IpAddresses.h

```diff
@@ -0,0 +1,551 @@
+		 * (it'll be ignored if subnet mask is correct)
+		 * @param[in] subnetMask The subnet mask to compare the address with the subnet
+		 */
+		#if __cplusplus > 201402L || _MSC_VER >= 1900
+		[[deprecated("This method is deprecated and will be deleted in the future. Prefer one of the functions matchSubnet")]]
+		#endif
+		bool matchSubnet(const IPv4Address& subnet, const IPv4Address& subnetMask) const;
+
+		/**
+		 * A static value representing a zero value of IPv4 address, meaning address of value "0.0.0.0"
+		 * Notice this value can be omitted in the user code because the default constructor creates an instance with an unspecified/zero address.
+		 * In order to check whether the address is zero the method isUnspecified can be used
+		 */
+		static const IPv4Address Zero;
+
+	private:
+		uint8_t m_Bytes[4];
+	}; // class IPv4Address
+
+
+	// Implementation of inline methods
+	uint32_t IPv4Address::toUInt() const
+	{
+		uint32_t addr;
+		memcpy(&addr, m_Bytes, sizeof(m_Bytes));
+		return addr;
+	}
+
+
+	// Creation of IPv4Address
+
+	/**
+	 * A function that creates an instance of the class IPv4Address out of 4-byte array.
+	 * @param[in] bytes The address as 4-byte array in network byte order
+	 * @return An instance of class IPv4Address
+	 */
+	inline IPv4Address makeIPv4Address(uint32_t addrAsInt) { return IPv4Address(addrAsInt); }
+
+	/**
+	 * A function that creates an instance of the class IPv4Address out of 4-byte array.
+	 * @param[in] bytes The address as 4-byte array in network byte order
+	 * @return An instance of class IPv4Address
+	 */
+	inline IPv4Address makeIPv4Address(const uint8_t bytes[4]) { return IPv4Address(bytes); }
+
+	/**
+	 * A function that creates an instance of the class out of string (char*) value
+	 * If the string doesn't represent a valid IPv4 address an errorCode will be set to non-zero value
+	 * @param[in] addressAsString The string (char*) representation of the address
+	 * @param[out] errorCode Contains 0 if a string represents a valid address, otherwise a non-zero value
+	 * @return An instance of class IPv4Address. If an error occured the return address will be an unspecified/zero
+	 */
+	IPv4Address makeIPv4Address(const char* addrAsString, int& errorCode);
+
+	/**
+	 * A function that creates an instance of the class out of std::string value
+	 * If the string doesn't represent a valid IPv4 address an errorCode will be set to non-zero value
+	 * @param[in] addressAsString The std::string representation of the address
+	 * @param[out] errorCode Contains 0 if a string represents a valid address, otherwise a non-zero value
+	 * @return An instance of class IPv4Address. If an error occured the return address will be an unspecified/zero
+	 */
+	inline IPv4Address makeIPv4Address(const std::string& addrAsString, int& errorCode) {	return makeIPv4Address(addrAsString.c_str(), errorCode); }
```

Comment:
> I guess the reason these methods are outside of the `IPv4Address` class (instead of having them as c'tors) is because of the error code they may return if the string is invalid. However the c'tor is very easy to use and much more intuitive for developers and that's why I implemented it that way.
> 
> If the user is not certain whether the IP address is valid they can always call the `isUnspecified()` method.

There are two ways to notify the user from constructor that an address is invalid:

1. make a flag `m_Valid` and method for checking (as done in current implementation)
2. throw an exception

The first approach increases the memory consumption and is inappropriate, especially for `IPv4Address` when the size grows twice. The second way is inappropriate because the library is not used the exceptions. So the static function is a decision. We always have an instance with a valid address.

In common case we cannot use `isUnspecified` to check whether the address is valid because an unspecified address is a valid. Yes, in the most cases an unspecified address can be considered as invalid.

Reply:
> BTW, I think that we can leave only the method that takes `std::string` and remove the one that takes `char *`. I've been waiting to do that for a long time and now seems to be a good opportunity 😃

Ok, let's remove it.

Reply:
> Another option to consider is to have them as static methods inside of `IPv4Address` class

If it's important I can make them static methods (and rename to `makeAddress`)

Reply:
I have made the function a static method

Reply:
in your implementation of this method the return value of an invalid string is an unspecified `IPv4Address` instance, so I'm thinking we basically get the same result if the user doesn't check the `errorCode` value.

The only reason I insist on having this discussion about a c'tor that takes a string is that it's much easier and intuitive for the users than any other solution (although it's less clean). I used these c'tors a lot throughout the code as you probably saw. Usually the user knows whether it's a valid IP address or not, and even if not he can check.

Another (quite ugly) solution we can think of is passing the `errorCode` parameter by reference to the c'tor. Again - it's not an ideal solution but it may serve both purposes of having easy-to-use code and one that is faster and consumes less memory.

Please let me know what you think

Reply:
Agree, I will make a constructor instead of method

Reply:
Done